### PR TITLE
feat(configureable-logger): add support of configurable logger.

### PR DESCRIPTION
**Pull request by haseeb-mhr**
## What
This PR Contains changes for adding configurable logger.

## Why
The reasons for the change is mentioned in this issue #80 .

## Type of change
Select multiple if applicable.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause a breaking change)
- [x] Tests (adds or updates tests)
- [ ] Documentation (adds or updates documentation)
- [ ] Refactor (style improvements, performance improvements, code refactoring)
- [ ] Revert (reverts a commit)
- [ ] CI/Build (adds or updates a script, change in external dependencies)

## Dependency Change
If a new dependency is being added, please ensure that it adheres to the following guideline https://github.com/apimatic/apimatic-codegen/wiki/Policy-of-adding-new-dependencies-in-the-core-libraries

## Breaking change
their is no breaking change in this PR. 

## Testing
Adding new test cases to test the changes.

## Checklist
- [ ] My code follows the coding conventions
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have added new unit tests

### CODE REVIEW

File: logger/sdkLogger.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+	if a.loggingOptions.request.includeQueryInPath {
+		url = request.RequestURI
+	} else {
+		url = a._removeQueryParams(request.RequestURI)
+	}
+
+	a.logger.Log(
+		logLevel,
+		fmt.Sprintf("Request %v %v %v", request.Method, url, contentTypeHeader),
+		map[string]any{
```
Comment (asadali214) [inline logger/sdkLogger.go:45]: @haseeb-mhr did Mujahid bhai went over this logging approach?

The purpose of structured logging is giving user a message format and parameters that they can replace in their message to get the real message.

Whereas you are replacing the message placeholders yourself before emitting the message to the loggerInterface^

File: logger/sdkLogger.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
+package logger
```
Comment (asadali214) [inline logger/sdkLogger.go:1]: This is a new package addition, but the core lib's readme is missing the documentation for the major classes in this namespace. Please add the documentation to the readme file

File: logger/sdkLogger.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	level Level,
+	request *http.Request,
+	logRequest RequestLoggerConfiguration) {
+
+	logHeaders := logRequest.headers
+	headersToInclude := logRequest.includeHeaders
+	headersToExclude := logRequest.excludeHeaders
+	headersToWhitelist := logRequest.whitelistHeaders
+
+	if logHeaders {
```
Comment (asadali214) [inline logger/sdkLogger.go:93]: `applyLogRequestHeaders` and `applyLogResponseHeaders` functions needs refactoring

They both are doing the same thing i.e getting a list of loggable headers from either request of response. You can simply create a GetLoggableHeaders(headers) function for the same thing

File: logger/sdkLogger.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+				filteredHeaders[key] = value
+			}
+		}
+	}
+	return filteredHeaders
+}
+
+func _contains(key string, slice []string) bool {
+	for _, name := range slice {
+		if name == key {
```
Comment (asadali214) [inline logger/sdkLogger.go:255]: found an instance of case sensitive header key matching. Headers matching should always be non sensitive

File: logger/sdkLogger.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+	return a._maskSenstiveHeaders(filteredHeaders, headersToWhitelist)
+}
+
+func (a *SdkLogger) _includeHeadersToLog(
+	headers, filteredHeaders http.Header,
+	headersToInclude []string) http.Header {
+	// Filter headers based on the keys specified in includeHeaders
+	for _, name := range headersToInclude {
+		val, ok := headers[name]
```
Comment (asadali214) [inline logger/sdkLogger.go:231]: found another instance of case sensitive header extraction. Headers extraction should always be non sensitive.

File: logger/sdkLogger.go (hunk 3 above)
Comment (haseeb-mhr) [reply logger/sdkLogger.go:93]: Refactor these function to implement log with minimal code

File: logger/sdkLogger.go (hunk 1 above)
Comment (haseeb-mhr) [reply logger/sdkLogger.go:45]: Updated this logic by implementing custom placeholder replacement.

File: logger/sdkLogger.go (hunk 5 above)
Comment (haseeb-mhr) [reply logger/sdkLogger.go:231]: updated the code to match non-sensitive headers matching

File: logger/sdkLogger.go (hunk 4 above)
Comment (haseeb-mhr) [reply logger/sdkLogger.go:255]: updated the code to match non-sensitive headers matching

File: logger/sdkLogger.go (hunk 2 above)
Comment (haseeb-mhr) [reply logger/sdkLogger.go:1]: added readme docs

File: logger/sdkLogger.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+	var contentLength string = ""
+	if len(headers) > 0 {
+		contentLength = headers.Get(CONTENT_LENGTH_HEADER)
+	}
+	return contentLength
+}
+
+func (a *SdkLogger) _removeQueryParams(url string) string {
+	if strIndex := strings.Index(url, "?"); strIndex != -1 {
+		return url[:strIndex]
```
Comment (asadali214) [inline logger/sdkLogger.go:149]: @haseeb-mhr You need to cover the uncovered lines in `sdkLogger.go` class

Comment (haseeb-mhr) [reply logger/sdkLogger.go:149]: updated request to improve code coverage.

**Review by asadali214 — DISMISSED** [review]

**Review by asadali214 — APPROVED** [review]

**Pull request merged**